pennylane_lightning/src/util/SparseLinAlg.hpp

```diff
@@ -160,16 +160,15 @@ void apply_Sparse_Matrix_Kokkos(
             });
         }
     }
+    return result;
```

Comment:
Is a return here better than passing in a pre-allocated buffer ?

Reply:
We used to have 
```
std::vector<std::complex<fp_precision>> apply_Sparse_Matrix
```
which declared `result` and then called `apply_Sparse_Matrix_Kokkos` (which `resize`s for robustness). To maintain the interface and functionality I just changed the return type of both `void` functions to `std::vector<std::complex<fp_precision>>` and renamed to `apply_Sparse_Matrix`. If we used this function in places where there is a preallocated buffer of the correct size, it would be better to pass it in and bypass `resize`, but I think it is not the case now.

Reply:
Thanks. I think as long as we can ensure we have RVO/copy elision (https://en.cppreference.com/w/cpp/language/copy_elision) with this, it should be fine. In this case, I suspect this design may be better capable of reasoning that, though it may become a problem if the SV is large, no?

Reply:
> In this case, I suspect this design may be better capable of reasoning that

I'm unsure about this. My intuition is most compilers would treat the previous and current designs the same and probably use RVO. 

> though it may become a problem if the SV is large, no?

You can't really do a general sparse matrix product in-place, so we (or Kokkos) need at least a temporary buffer the size of SV. I think the current design chooses not to keep this buffer around to save memory. I guess we can run many circuits without that buffer, but we could accelerate (likely not by much) some circuits such as those with sparse Hamiltonians.